### ISSUE TITLE
[mlir] -remove-dead-values generates invalid MLIR for affine.for with iter_args, which executes only once.

### ISSUE THREAD

**Issue by yyong120** [llvm/llvm-project#134291]
To reproduce: `mlir-opt -remove-dead-values test.mlir`

`test.mlir`:
```
func.func @test_affine_for() {
    %c1_i32 = arith.constant 1 : i32
    %alloca = memref.alloca() : memref<16xi32>
    %0 = affine.for %arg0 = 0 to 16 step 64 iter_args(%arg1 = %c1_i32) -> (i32) {
        %1 = arith.addi %arg1, %c1_i32 : i32
        memref.store %1, %alloca[%arg0] : memref<16xi32>
        affine.yield %1 : i32
    }
    return
}
```

The generated `result.mlir`:
```
func.func @test_for() {
    %c1_i32 = arith.constant 1 : i32
    %alloca = memref.alloca() : memref<16xi32>
    affine.for %arg0 = 0 to 16 step 64 iter_args(%arg1 = %c1_i32) -> () {
        %0 = arith.addi %arg1, %c1_i32 : i32
        memref.store %0, %alloca[%arg0] : memref<16xi32>
        affine.yield
    }
    return
}
```
Run `mlir-opt --verify-diagnostics result.mlir`, the error is

```
result_1.mlir:4:5: error: unexpected error: custom op 'affine.for' mismatch between the number of loop-carried values and results
    affine.for %arg0 = 0 to 16 step 64 iter_args(%arg1 = %c1_i32) -> () {
    ^
```

This issue happens when an `affine.for` only executes once. Since the loop-carried arguments (`iter_args`) will not be updated, remove-dead-values pass will remove the return values of `affine.yield`. However, `iter_args` are still left in the for loop, which leads to an invalid mlir.

**Comment by Baxi-codes** [comment]
Hi, should it remove the iter_args altogether or leave it intact?

**Comment by yyong120** [comment]
> Hi, should it remove the iter_args altogether or leave it intact?

I think either way is fine. If we decide to *"remove the iter_args altogether"*, they should be defined outside the `affine.for` loop.

**Comment by ed741** [comment]
Hi,
It appears that `-remove-dead-values` is removing the operand from `yield` because `affine.for` does not include it's body region as a successor of the body region if it is known to only do one iteration. So the remove dead values transform does not see that the operand should be maintained for the sake of the (never-going-to-happen) next iteration.

The parser for `affine.for` enforces that the number of block arguments (minus the index variable) and results be the same which is not then the case when the result and yield operand are removed.
But the verifier `AffineForOp::verifyRegions()` assumes success if the number or results of the op is 0, and so doesn't check if the number or type of iteration arguments lines up.

So if `affine.for` does not consider the body region to be successor of itself when there is one iteration, should it really be verifying & parsing as if it is a successor? Or the implementation of `RegionBranchOpInterface` for `affine.for` so it doesn't omit the body as a successor of itself.
